### PR TITLE
Improve CPU node placement logic to avoid Memcpys

### PR DESCRIPTION
**Pull request by ashbhandare**
The current logic for CPU node placement for small shape-related computations goes top-down tracing from a node that produces the output on CPU, and stops at the first node that consumes a CPU input while being able to execute on GPU. 
However, if there are further downstream nodes that also accept CPU input while executing on GPU, a D2H memcpy happens in this case. For example:

This PR modifies the logic to follow the forward trace with a backward trace, which looks for more ops that can be placed on CPU, while stopping till the trace either hits a cpu node, or a node that produces and output on CPU.
The following test graphs are added as well:
`cpu_fallback_pattern_0.onnx`
cpu_fallback_pattern_0.onnx
`cpu_fallback_pattern_1.onnx`
cpu_fallback_pattern_1.onnx
`cpu_fallback_pattern_2.onnx`
cpu_fallback_pattern_2.onnx
`cpu_fallback_pattern_3.onnx`
cpu_fallback_pattern_3.onnx
`cpu_fallback_pattern_4.onnx`
cpu_fallback_pattern_4.onnx
Topological ordering is used to trace in both directions, as covered by the following test:
`cpu_fallback_pattern_5.onnx`
cpu_fallback_pattern_5.onnx

### CODE REVIEW

File: onnxruntime/test/framework/session_state_test.cc [hunk 1]
```diff
@@ -173,6 +179,72 @@ TEST_P(SessionStateTestP, TestInitializerProcessing) {
   }
 }
 
+#if defined(USE_CUDA) || defined(USE_ROCM)
+static void TestCPUNodePlacement(const std::basic_string<ORTCHAR_T>& model_uri, size_t expected_cpu_cnt, size_t expected_gpu_cnt) {
```
Comment (SherlockNoMad) [inline onnxruntime/test/framework/session_state_test.cc:183]: Good test!

File: onnxruntime/test/framework/session_state_test.cc [hunk 2]
```diff
@@ -173,6 +179,72 @@ TEST_P(SessionStateTestP, TestInitializerProcessing) {
...
+    auto& ep = node.GetExecutionProviderType();
+    if (ep == onnxruntime::kCudaExecutionProvider || ep == onnxruntime::kRocmExecutionProvider) {
+      ++actual_gpu_cnt;
+    } else if (ep == onnxruntime::kCpuExecutionProvider) {
+      ++actual_cpu_cnt;
+    } else {
+      ASSERT_TRUE(false) << "Invalid execution provider assigned to node: " << node.Name() << " , value: " << ep;
+    }
+  }
+  ASSERT_EQ(expected_cpu_cnt, actual_cpu_cnt);
```
Comment (SherlockNoMad) [inline onnxruntime/test/framework/session_state_test.cc:227]: Can we assert on the expected CPU node name directly? 
I know the node count is widely used in graph transformer UTs, but I am not the biggest fan of this approach. 
We can do more explicit assert than counting node. :)

File: onnxruntime/core/framework/fallback_cpu_capability.cc [hunk 3]
```diff
@@ -71,11 +77,26 @@ std::unordered_set<NodeIndex> GetCpuPreferredNodes(const onnxruntime::GraphViewe
...
+              LOGS_DEFAULT(INFO) << "Candidate for fallback CPU execution in forward trace: " << consumer_node->Name();
+            }
+          }
+          return Status::OK();
+        }));
+
+    // then, find all the direct producers of cpu tensors.
+    ORT_THROW_IF_ERROR(node->ForEachWithIndex(
+        node->InputDefs(),
+        [&](const NodeArg& node_arg, size_t ip_index) {
```
Comment (SherlockNoMad) [inline onnxruntime/core/framework/fallback_cpu_capability.cc:93]: in_index?

File: onnxruntime/core/framework/fallback_cpu_capability.cc [hunk 4]
```diff
@@ -136,10 +157,69 @@ std::unordered_set<NodeIndex> GetCpuPreferredNodes(const onnxruntime::GraphViewe
...
-        candidates.push((*it).Index());
+        candidates_fw.push((*it).Index());
+      }
+    }
+  }
+  // clear the visited to prepare for backward trace
+  visited.clear();
+  // Trace the graph backwards to find additional CPU nodes
+  // Starting from nodes that must produce an output on CPU, trace the producer nodes
+  // The trace stops when 1) we find that the node is already picked for CPU
```
Comment (SherlockNoMad) [inline onnxruntime/core/framework/fallback_cpu_capability.cc:171]: new line for each bullet point 
1)
2)
3)
4)

File: onnxruntime/core/framework/fallback_cpu_capability.cc [hunk 5]
```diff
@@ -45,10 +45,16 @@ std::unordered_set<NodeIndex> GetCpuPreferredNodes(const onnxruntime::GraphViewe
...
   };
 
-  std::priority_queue<NodeIndex, std::vector<NodeIndex>, decltype(greater_order_comp)> candidates(greater_order_comp);
+  // If return false, n2 will be output first; If return true, n1 will be output first
+  auto lesser_order_comp = [&](const NodeIndex n1, const NodeIndex n2) {
+    return node_id_to_order_map[n1] < node_id_to_order_map[n2];
+  };
+
+  std::priority_queue<NodeIndex, std::vector<NodeIndex>, decltype(greater_order_comp)> candidates_fw(greater_order_comp);
+  std::priority_queue<NodeIndex, std::vector<NodeIndex>, decltype(lesser_order_comp)> candidates_bw(lesser_order_comp);
```
Comment (SherlockNoMad) [inline onnxruntime/core/framework/fallback_cpu_capability.cc:54]: There is an implicit assumption here: NodeIndex is an approximate ordering for the execution order. 
I think this is a wrong assumption. 
To get the real execution order, we need to do an topological sort. 
Let me construct an offending case for you.

Comment (SherlockNoMad) [reply onnxruntime/core/framework/fallback_cpu_capability.cc:54]: image

Comment (SherlockNoMad) [reply onnxruntime/core/framework/fallback_cpu_capability.cc:54]: In the forward pass, by the time we visit Concat node, the right input is not yet placed in cpu_args yet... 
As a result, concat node would not be placed on CPU.

Comment (SherlockNoMad) [reply onnxruntime/core/framework/fallback_cpu_capability.cc:54]: Let cover such cases of  "multi-starting point, multi-ending point" in the UT.

Comment (SherlockNoMad) [reply onnxruntime/core/framework/fallback_cpu_capability.cc:54]: This is something I learnt during the Cast Propagation work. 
The algorithm we are we are taking here is "CPU colored-subgraph" propagation. 
And to propagate the "frontier" of a colored-subgraph, we should always use topological order.

Comment (ashbhandare) [reply onnxruntime/core/framework/fallback_cpu_capability.cc:57]: Topological ordering is already in use in the comparator function : https://github.com/microsoft/onnxruntime/blob/331cd1c824b885447b13ea8656d2674e51b002b3/onnxruntime/core/framework/fallback_cpu_capability.cc#L36
I have added the test you specified as well.

File: onnxruntime/core/framework/fallback_cpu_capability.cc [hunk 6]
```diff
@@ -57,12 +60,29 @@ std::unordered_set<NodeIndex> GetCpuPreferredNodes(const onnxruntime::GraphViewe
...
 
     const KernelCreateInfo* kernel_info = nullptr;
+
+    // Get the CPU kernel availability for this node
+    KernelRegistryManager mgr;
+    ExecutionProviders cpu_ep;
+    CPUExecutionProviderInfo epi{false};
+    ORT_ENFORCE(cpu_ep.Add(kCpuExecutionProvider, onnxruntime::make_unique<CPUExecutionProvider>(epi)).IsOK());
+    ORT_ENFORCE(mgr.RegisterKernels(cpu_ep).IsOK());
+    std::vector<const KernelRegistry*> cpu_kernel_registries = mgr.GetKernelRegistriesByProviderType(kCpuExecutionProvider);
```
Comment (SherlockNoMad) [inline onnxruntime/core/framework/fallback_cpu_capability.cc:77]: This kernel registry should be built once outside the loop. 
ORT has an implicit rule that all the ops would have an CPU kernel, so I am wondering if this check is necessary.

Comment (SherlockNoMad) [reply onnxruntime/core/framework/fallback_cpu_capability.cc:77]: Did you encounter a case where a node is placed on CPU, but the cpu kernel is missing?

Comment (ashbhandare) [reply onnxruntime/core/framework/fallback_cpu_capability.cc:77]: We don't have CPU kernels for some contrib ops like IsAllFinite. Some tests in ORTTrainer were failing as the output of IsAllFinite was consumed by AdamOptiomizer/LambOptimizer, which bind the "do_update" input to CPU which is produced by "IsAllFinite".

**Review by SherlockNoMad — APPROVED** [review]

**Review by SherlockNoMad — APPROVED** [review]

**Pull request merged**